src/rpc/rawtransaction.cpp

```diff
@@ -1745,101 +1537,296 @@ UniValue signrawtransaction(const UniValue& params, bool fHelp)
+                                  MutableTransactionSignatureChecker(&mergedTx, nAllInputsIndex), &serror))
+                {
+                    TxCswInErrorToJSON(txCswIn, i, vErrors, ScriptErrorString(serror));
+                }
+            }
+        }
 
-            const CScript& prevPubKey = txCswIn.scriptPubKey();
+        bool fComplete = vErrors.empty();
 
-            txCswIn.redeemScript.clear();
-            // Only sign SIGHASH_SINGLE if there's a corresponding output:
-            // Note: we should consider the regular inputs as well.
-            if (!fHashSingle || (nAllInputsIndex < mergedTx.getVout().size()))
-                SignSignature(keystore, prevPubKey, mergedTx, nAllInputsIndex, nHashType);
+        UniValue result(UniValue::VOBJ);
+        result.pushKV("hex", EncodeHexTx(CTransaction(mergedTx)));
+        result.pushKV("complete", fComplete);
+        if (!vErrors.empty()) {
+            result.pushKV("errors", vErrors);
+        }
 
-            // ... and merge in other signatures:
-            /* Note:
-             * For CTxCeasedSidechainWithdrawalInput currently only P2PKH is allowed.
-             * SignSignature can return true and set `txCswIn.redeemScript` value in case there is a proper private key in the keystore.
-             * It can return false and leave `txCswIn.redeemScript` empty in case of any error occurs.
-             * CombineSignatures will try to get the most recent signature:
-             * 1) if SignSignature operation was successful -> leave `txCswIn.redeemScript value as is.
-             * 2) if SignSignature operation was unsuccessful -> set `txCswIn.redeemScript value equal to the origin `txv` csw input script.
-             * Later the signature will be checked, so in case no origin signature and no new one exist -> verification will fail.
-             */
-            for(const CMutableTransaction& txv : txVariants)
-                txCswIn.redeemScript = CombineSignatures(prevPubKey, mergedTx, nAllInputsIndex, txCswIn.redeemScript, txv.vcsw_ccin[i].redeemScript);
+        return result;
+    } else {
+        // Fetch previous transactions (inputs):
+        CCoinsView viewDummy;
+        CCoinsViewCache view(&viewDummy);
+        {
+            LOCK(mempool.cs);
+            CCoinsViewCache &viewChain = *pcoinsTip;
+            CCoinsViewMemPool viewMempool(&viewChain, mempool);
+            view.SetBackend(viewMempool); // temporarily switch cache backend to db+mempool view
+
+            BOOST_FOREACH(const CTxIn& txin, certificate.vin) {
+                const uint256& prevHash = txin.prevout.hash;
+                CCoins coins;
+                view.AccessCoins(prevHash); // this is certainly allowed to fail
+            }
+
+            view.SetBackend(viewDummy); // switch back to avoid locking mempool for too long
+        }
+
+        // Add previous txouts given in the RPC call:
+        if (params.size() > 1 && !params[1].isNull()) {
```

Comment:
This block is duplicated for both Tx and Cert. Move it back outside as it was before.